# Práctica 1 - Tipos Numéricos, String, Set, Diccionarios, Tuplas

preparado por prof. Antonio Russoniello

## Paso 1

Vamos a crear la variable **mensaje** esta variable contiene valores ambientales de humedad y temperatura para varias ciudades del país

In [1]:
mensaje = 'Valencia->44%,24.385417062434136;Valencia->41%,23.16706208744312;Maracay->54%,23.174235266004455;'\
        "Caracas->60%,23.21881955940879;Maracaibo->46%,23.08528369893707;Amazonas->59%,24.153870621406824;"\
        'Amazonas->46%,24.112679522556327;Caracas->50%,23.033279906706532;Caracas->43%,24.471850403219218;'\
        'Caracas->60%,23.093638109777544;Amazonas->40%,23.27146416378274;Caracas->44%,24.136596515560008;'\
        'Caracas->41%,23.211896444367312;Amazonas->60%,24.306861635724474;Caracas->56%,23.24969618699264;'\
        'Caracas->54%,23.18420581435151;Caracas->51%,24.451987950007332;Maracaibo->45%,24.44897952121968;'\
        'Caracas->55%,24.237886442373526;Maracaibo->60%,23.12115843953649;Valencia->43%,24.17915318123484;'\
        'Valencia->40%,23.294380429607685;Valencia->54%,23.025011717022654;Maracay->42%,23.413523250262084;'\
        'Maracay->44%,23.1614712307336'

queremos crear una lista que contenga cada elemento de la lista **mensaje**, para ello procedemos a separar los elementos por el caracter **;**

In [2]:
data_ciudad = mensaje.split(';')

presentemos el valor de la variable **data_ciudad** de tipo "String"

In [3]:
print(data_ciudad)

['Valencia->44%,24.385417062434136', 'Valencia->41%,23.16706208744312', 'Maracay->54%,23.174235266004455', 'Caracas->60%,23.21881955940879', 'Maracaibo->46%,23.08528369893707', 'Amazonas->59%,24.153870621406824', 'Amazonas->46%,24.112679522556327', 'Caracas->50%,23.033279906706532', 'Caracas->43%,24.471850403219218', 'Caracas->60%,23.093638109777544', 'Amazonas->40%,23.27146416378274', 'Caracas->44%,24.136596515560008', 'Caracas->41%,23.211896444367312', 'Amazonas->60%,24.306861635724474', 'Caracas->56%,23.24969618699264', 'Caracas->54%,23.18420581435151', 'Caracas->51%,24.451987950007332', 'Maracaibo->45%,24.44897952121968', 'Caracas->55%,24.237886442373526', 'Maracaibo->60%,23.12115843953649', 'Valencia->43%,24.17915318123484', 'Valencia->40%,23.294380429607685', 'Valencia->54%,23.025011717022654', 'Maracay->42%,23.413523250262084', 'Maracay->44%,23.1614712307336']


presentemos por pantalla cada elemento de la lista **data_ciudad**

In [4]:
for data in data_ciudad:
    print(data)

Valencia->44%,24.385417062434136
Valencia->41%,23.16706208744312
Maracay->54%,23.174235266004455
Caracas->60%,23.21881955940879
Maracaibo->46%,23.08528369893707
Amazonas->59%,24.153870621406824
Amazonas->46%,24.112679522556327
Caracas->50%,23.033279906706532
Caracas->43%,24.471850403219218
Caracas->60%,23.093638109777544
Amazonas->40%,23.27146416378274
Caracas->44%,24.136596515560008
Caracas->41%,23.211896444367312
Amazonas->60%,24.306861635724474
Caracas->56%,23.24969618699264
Caracas->54%,23.18420581435151
Caracas->51%,24.451987950007332
Maracaibo->45%,24.44897952121968
Caracas->55%,24.237886442373526
Maracaibo->60%,23.12115843953649
Valencia->43%,24.17915318123484
Valencia->40%,23.294380429607685
Valencia->54%,23.025011717022654
Maracay->42%,23.413523250262084
Maracay->44%,23.1614712307336


## Paso 2

Ahora queremos que los datos de **data_ciudad** sean asignados a una variable de tipo lista que contenga como elementos la lista de valores, debe convertirse la temperatura a tipo flotante y presentarse con dos dígitos decimales.

Se presenta un ejemplo de la salida deseada: <br>

['Valencia', '44%', 24.39],
['Valencia', '41%', 23.17],
['Maracay', '54%', 23.17],
['Caracas', '60%', 23.22]
...
]

In [5]:
nueva_lista = []

for data in data_ciudad:
    elemento = data.split(',')
    elemento_tmp = elemento[0].split('->')
    ciudad = elemento_tmp[0]
    hum = elemento_tmp[1]
    temp = round(float(elemento[1]), 2)    
    valores = [ciudad,hum,temp]
    nueva_lista.append(valores)

print(nueva_lista)

[['Valencia', '44%', 24.39], ['Valencia', '41%', 23.17], ['Maracay', '54%', 23.17], ['Caracas', '60%', 23.22], ['Maracaibo', '46%', 23.09], ['Amazonas', '59%', 24.15], ['Amazonas', '46%', 24.11], ['Caracas', '50%', 23.03], ['Caracas', '43%', 24.47], ['Caracas', '60%', 23.09], ['Amazonas', '40%', 23.27], ['Caracas', '44%', 24.14], ['Caracas', '41%', 23.21], ['Amazonas', '60%', 24.31], ['Caracas', '56%', 23.25], ['Caracas', '54%', 23.18], ['Caracas', '51%', 24.45], ['Maracaibo', '45%', 24.45], ['Caracas', '55%', 24.24], ['Maracaibo', '60%', 23.12], ['Valencia', '43%', 24.18], ['Valencia', '40%', 23.29], ['Valencia', '54%', 23.03], ['Maracay', '42%', 23.41], ['Maracay', '44%', 23.16]]


presente por pantalla cada elemento de la lista **nueva_lista** con el siguiente formato:<br>

ciudad: Valencia humedad: 44% temperatura: 24.39'


In [6]:
datos_ciudad = []
for data in nueva_lista:
    ciudad = data[0]
    hum = data[1]
    temp = data[2]
    print('ciudad: ' + ciudad + ' humedad: ' + hum + ' temperatura: ' + str(temp))

ciudad: Valencia humedad: 44% temperatura: 24.39
ciudad: Valencia humedad: 41% temperatura: 23.17
ciudad: Maracay humedad: 54% temperatura: 23.17
ciudad: Caracas humedad: 60% temperatura: 23.22
ciudad: Maracaibo humedad: 46% temperatura: 23.09
ciudad: Amazonas humedad: 59% temperatura: 24.15
ciudad: Amazonas humedad: 46% temperatura: 24.11
ciudad: Caracas humedad: 50% temperatura: 23.03
ciudad: Caracas humedad: 43% temperatura: 24.47
ciudad: Caracas humedad: 60% temperatura: 23.09
ciudad: Amazonas humedad: 40% temperatura: 23.27
ciudad: Caracas humedad: 44% temperatura: 24.14
ciudad: Caracas humedad: 41% temperatura: 23.21
ciudad: Amazonas humedad: 60% temperatura: 24.31
ciudad: Caracas humedad: 56% temperatura: 23.25
ciudad: Caracas humedad: 54% temperatura: 23.18
ciudad: Caracas humedad: 51% temperatura: 24.45
ciudad: Maracaibo humedad: 45% temperatura: 24.45
ciudad: Caracas humedad: 55% temperatura: 24.24
ciudad: Maracaibo humedad: 60% temperatura: 23.12
ciudad: Valencia humedad: 43

## Paso 3

ahora requerimos reasignar los valores a una nueva lista de manera que podamos ordenarla por orden alfabético de ciudad y obtener por pantalla los valores agrupados. Deben ser presentados con el siguiente formato:<br>
<img src="salida.png" alt="Alternative text" />
<center>...

In [7]:
datos_ciudad = []
for data in data_ciudad:
    elemento = data.split(',')
    elemento_tmp = elemento[0].split('->')
    ciudad = elemento_tmp[0]
    hum = elemento_tmp[1]
    temp = elemento[1]
    datos_ciudad.append(ciudad +  ' ' + hum + ' ' + str(temp))
datos_ciudad.sort()
print(datos_ciudad)

['Amazonas 40% 23.27146416378274', 'Amazonas 46% 24.112679522556327', 'Amazonas 59% 24.153870621406824', 'Amazonas 60% 24.306861635724474', 'Caracas 41% 23.211896444367312', 'Caracas 43% 24.471850403219218', 'Caracas 44% 24.136596515560008', 'Caracas 50% 23.033279906706532', 'Caracas 51% 24.451987950007332', 'Caracas 54% 23.18420581435151', 'Caracas 55% 24.237886442373526', 'Caracas 56% 23.24969618699264', 'Caracas 60% 23.093638109777544', 'Caracas 60% 23.21881955940879', 'Maracaibo 45% 24.44897952121968', 'Maracaibo 46% 23.08528369893707', 'Maracaibo 60% 23.12115843953649', 'Maracay 42% 23.413523250262084', 'Maracay 44% 23.1614712307336', 'Maracay 54% 23.174235266004455', 'Valencia 40% 23.294380429607685', 'Valencia 41% 23.16706208744312', 'Valencia 43% 24.17915318123484', 'Valencia 44% 24.385417062434136', 'Valencia 54% 23.025011717022654']


In [8]:
c = ""
for ciudad in datos_ciudad:
    parse_ciudad = ciudad.split(' ')
    if parse_ciudad[0] != c:
        c = parse_ciudad[0]                 
        print('*'*40)                 
        print('datos para: ' + c)
        print('-'*40)               
        print('humedad\t\ttemperatura')
        print('-'*40)                
    print(parse_ciudad[1] + '\t\t' + parse_ciudad[2])

****************************************
datos para: Amazonas
----------------------------------------
humedad		temperatura
----------------------------------------
40%		23.27146416378274
46%		24.112679522556327
59%		24.153870621406824
60%		24.306861635724474
****************************************
datos para: Caracas
----------------------------------------
humedad		temperatura
----------------------------------------
41%		23.211896444367312
43%		24.471850403219218
44%		24.136596515560008
50%		23.033279906706532
51%		24.451987950007332
54%		23.18420581435151
55%		24.237886442373526
56%		23.24969618699264
60%		23.093638109777544
60%		23.21881955940879
****************************************
datos para: Maracaibo
----------------------------------------
humedad		temperatura
----------------------------------------
45%		24.44897952121968
46%		23.08528369893707
60%		23.12115843953649
****************************************
datos para: Maracay
----------------------------------------
hu

## Paso 4

En este paso asignaremos a una variable de tipo diccionario una lista con el promedio de valor de húmedad y temperatura para cada ciudad, la ciudad será usada como clave del diccionario

In [9]:
lista_ciudades = []

for ciudad in nueva_lista:
    lista_ciudades.append(ciudad[0])

conj_ciudades=set(lista_ciudades) # podemos usar una variable tipo Set para crear un iterable de ciudad

valores_prom = {}
for elementos in conj_ciudades:
    valores_prom[elementos] = [] # se crea un diccionario que contiene listas como valores

valores_prom # mostramos la estructura de la variable tipo diccionario

{'Amazonas': [], 'Maracaibo': [], 'Caracas': [], 'Maracay': [], 'Valencia': []}

vamos a visualizar la lista **nueva_lista** trabajaremos con ella para extraer los datos requeridos

In [10]:
print(nueva_lista)

[['Valencia', '44%', 24.39], ['Valencia', '41%', 23.17], ['Maracay', '54%', 23.17], ['Caracas', '60%', 23.22], ['Maracaibo', '46%', 23.09], ['Amazonas', '59%', 24.15], ['Amazonas', '46%', 24.11], ['Caracas', '50%', 23.03], ['Caracas', '43%', 24.47], ['Caracas', '60%', 23.09], ['Amazonas', '40%', 23.27], ['Caracas', '44%', 24.14], ['Caracas', '41%', 23.21], ['Amazonas', '60%', 24.31], ['Caracas', '56%', 23.25], ['Caracas', '54%', 23.18], ['Caracas', '51%', 24.45], ['Maracaibo', '45%', 24.45], ['Caracas', '55%', 24.24], ['Maracaibo', '60%', 23.12], ['Valencia', '43%', 24.18], ['Valencia', '40%', 23.29], ['Valencia', '54%', 23.03], ['Maracay', '42%', 23.41], ['Maracay', '44%', 23.16]]


Procedemos a crear nuestro diccionario **valores_prom**

In [11]:
for ciudad in conj_ciudades:
    hum, temp, contador = 0, 0, 0
    for elemento in nueva_lista:
        if ciudad == elemento[0]:
            contador+=1
            hum += int(elemento[1][:-1])
            temp += elemento[2]
            valores_prom[ciudad]= [round(hum/contador,2),round(temp/contador,2)]
        
valores_prom

{'Amazonas': [51.25, 23.96],
 'Maracaibo': [50.33, 23.55],
 'Caracas': [51.4, 23.63],
 'Maracay': [46.67, 23.25],
 'Valencia': [44.4, 23.61]}